patches/v8/0012-Implement-Promise-Context-Tagging.patch

```diff
@@ -0,0 +1,468 @@
+ include/v8-isolate.h                        | 16 +++++++++
+ src/api/api.cc                              | 17 ++++++++++
+ src/builtins/promise-abstract-operations.tq | 11 ++++--
+ src/builtins/promise-constructor.tq         |  4 +++
+ src/builtins/promise-misc.tq                |  4 +++
+ src/compiler/js-create-lowering.cc          |  4 ++-
+ src/diagnostics/objects-printer.cc          |  1 +
+ src/execution/isolate-inl.h                 | 20 +++++++++++
+ src/execution/isolate.cc                    | 37 +++++++++++++++++++++
+ src/execution/isolate.h                     | 15 +++++++++
+ src/heap/factory.cc                         |  6 ++++
+ src/objects/js-promise.tq                   |  1 +
+ src/profiler/heap-snapshot-generator.cc     |  3 ++
+ src/runtime/runtime-promise.cc              | 35 +++++++++++++++++++
+ src/runtime/runtime.h                       |  4 ++-
+ 16 files changed, 182 insertions(+), 5 deletions(-)
+
+diff --git a/include/v8-callbacks.h b/include/v8-callbacks.h
+index 2743d5cc69e3c299bb3b84e1e97424b4dee0ae9e..b862ead1533eccfc9340ae6031c1bdab301550a8 100644
+--- a/include/v8-callbacks.h
++++ b/include/v8-callbacks.h
+@@ -413,6 +413,15 @@ using PrepareStackTraceCallback = MaybeLocal<Value> (*)(Local<Context> context,
+                                                         Local<Value> error,
+                                                         Local<Array> sites);
+ 
++/**
++ * PromiseCrossContextCallback is called when following a promise and the
++ * promise's context tag is not strictly equal to the isolate's current
++ * promise context tag.
++ */
++using PromiseCrossContextCallback = MaybeLocal<Promise> (*)(Local<Context> context,
++                                                            Local<Promise> promise,
++                                                            Local<Object> tag);
++
+ }  // namespace v8
+ 
+ #endif  // INCLUDE_V8_ISOLATE_CALLBACKS_H_
+diff --git a/include/v8-isolate.h b/include/v8-isolate.h
+index 1d079ac0bb1dd3921d680bcd1958da4bbc114e9e..4d1ab8bd462359a8173cdce6e8173e475f879690 100644
+--- a/include/v8-isolate.h
++++ b/include/v8-isolate.h
+@@ -1635,6 +1635,9 @@ class V8_EXPORT Isolate {
+    */
+   void LocaleConfigurationChangeNotification();
+ 
++  class PromiseContextScope;
++  void SetPromiseCrossContextCallback(PromiseCrossContextCallback callback);
++
+   Isolate() = delete;
+   ~Isolate() = delete;
+   Isolate(const Isolate&) = delete;
+@@ -1678,6 +1681,19 @@ MaybeLocal<T> Isolate::GetDataFromSnapshotOnce(size_t index) {
+   return Local<T>::FromSlot(slot);
+ }
+ 
++class Isolate::PromiseContextScope {
++public:
++  PromiseContextScope(Isolate* isolate, v8::Local<v8::Object> tag);
++  ~PromiseContextScope();
++  PromiseContextScope(const PromiseContextScope&) = delete;
++  PromiseContextScope(PromiseContextScope&&) = delete;
++  PromiseContextScope& operator=(const PromiseContextScope&) = delete;
++  PromiseContextScope& operator=(PromiseContextScope&&) = delete;
```

Comment:
V8 would typically use `DISALLOW_IMPLICIT_CONSTRUCTORS(PromiseContextScope)` for all but the move constructor here (not fanaticially used everywhere, jfyi).

Reply:
Overall this patch LGTM. I've run v8 tests against the new patch 0012 for release and debug builds with no issues.